### PR TITLE
chore(federation): Improved serialization impls

### PR DESCRIPTION
**Pull request by TylerBloom**
This PR changes how several types are serialized in `apollo-federation`. Namely, there are several types from `apollo-compiler` that do not implement `serde::Serialize`. Before, these types were serialized as strings or simply skipped. This has been changes to use the provided `serialize` method on those `apollo-compiler` types. Additionally, this changes the inner type of `SelectionId` and `OverrideId` from a `usize` to a `Uuid`. This adds `uuid` as a dependency to `apollo-federation`, but `uuid` was already a (transitive) dependency of `apollo-router`.

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[ROUTER-841]: https://apollographql.atlassian.net/browse/ROUTER-841?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apollo-federation/src/operation/mod.rs [hunk 1]
```diff
@@ -66,23 +65,13 @@ pub(crate) use rebase::*;
...
-///
-/// Note that we shouldn't add `derive(Serialize, Deserialize)` to this without changing the types
-/// to be something like UUIDs.
-#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
-// NOTE(@TylerBloom): This feature gate can be removed once the condition in the comment above is
-// met. Note that there are `serde(skip)` statements that should be removed once this is removed.
-#[cfg_attr(feature = "snapshot_tracing", derive(Serialize))]
-pub(crate) struct SelectionId(usize);
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash, serde::Serialize)]
+pub(crate) struct SelectionId(Uuid);
```
Comment (dariuszkuc) [inline apollo-federation/src/operation/mod.rs:70]: should this change be in a separate PR?

**Comment by goto-bus-stop** [comment]
I'd prefer not to introduce UUID if we can avoid it, I like the rest though 😁

File: apollo-federation/src/operation/mod.rs (hunk 1 above)
Comment (TylerBloom) [reply apollo-federation/src/operation/mod.rs:70]: This can be moved into a different PR or reverted. My motivation was to make our serialization uniform (i.e. not feature gating anything), and both using `apollo-compiler`'s `serialize` methods and removing feature-gating fall into this. While we aren't serializing to cache, simply implementing `Serialize` would violate what that comment is saying but changing the representation and not cache is still inline.

My goal with making things uniform (and specifically removing the feature-gating) was to minimize what we would need to consider when/if we started to serialize for caching. That said, this is ultimately a small change that can be revisited at that point in time if this change isn't inline with what would be needed.

Comment (goto-bus-stop) [reply apollo-federation/src/operation/mod.rs:70]: I think the comment is a bit over-eager and it should maybe just generally warn about it being a fleeting value that is not designed to be unique between planning runs

Comment (goto-bus-stop) [reply apollo-federation/src/operation/mod.rs:70]: Oh, the uniform attributes point makes sense. I think it's okay to not feature gate serialization and warn about deserialization in the comment?

Comment (TylerBloom) [reply apollo-federation/src/operation/mod.rs:70]: >  I think it's okay to not feature gate serialization and warn about deserialization in the comment?

I think this works. We already do not implement `Deserialize` and there are several outstanding questions around what that would look like (there is even a new one in the PR). I'll revert the `uuid` changes, update that comment, and leave a new comment (either at the top of `lib.rs` or by `QueryPlan`) saying that we don't support deserialization yet and we need to answer a few questions before we can.

**Review by dariuszkuc — APPROVED** [review]
